File: javascript/linkedlist/ll.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const node_1 = __importDefault(require("./node"));
class LinkedList {
    insertFirst(val) {
        let node = new node_1.default(val);
        node.next = this.head;
        this.head = node;
        if (this.tail == null) {
            this.tail = this.head;
        }
        this.size++;
    }
    insertLast(val) {
        if (this.tail == null) {
            this.insertFirst(val);
            return;
        }
        let node = new node_1.default(val);
        this.tail.next = node;
        this.tail = node;
        this.size++;
    }
    // insert at index
    insertAt(val, index) {
        if (index == 0) {
            this.insertFirst(val);
            return;
        }
        if (index == this.size) {
            this.insertLast(val);
            return;
        }
        let node = new node_1.default(val);
        let temp = this.get(index - 1);
        node.next = temp.next;
        temp.next = node;
        this.size++;
    }
    deleteFirst() {
        let val = this.head.val;
        if (this.head == null) {
            return;
        }
        this.head = this.head.next;
        this.size--;
        return val;
    }
    deleteLast() {
        let val = this.tail.val;
        if (this.size <= 1) {
            this.deleteFirst();
        }
        let secondlast = this.get(this.size - 2);
        this.tail = secondlast;
        this.tail.next = null;
        return val;
    }
    deleteAt(index) {
        if (index == 0) {
            this.deleteFirst();
        }
        if (index === this.size - 1) {
            this.deleteLast();
        }
        let prev = this.get(index - 1);
        let val = prev.next.val;
        prev.next = prev.next.next;
        this.size--;
        return val;
    }
    get(index) {
        let temp = this.head;
        for (let i = 0; i < index; i++) {
            temp = temp.next;
        }
        return temp;
    }
    display() {
        let temp = this.head;
        while (temp != null) {
            console.log(temp.val);
            console.log("|");
            temp = temp.next;
        }
        console.log("null");
    }
}
let list = new LinkedList();
list.insertFirst(15);
list.insertFirst(11);
list.insertFirst(8);
list.insertFirst(5);
list.insertLast(100);
list.insertAt(69, 0);
list.display();

File: golang/linkedlist/LL.go
package main

import (
	"fmt"
)

type Node struct {
	data int
	next *Node
}

type LinkedList struct {
	head *Node
	tail *Node
	size int
}

func (l *LinkedList) insertFirst(data int) {
	node := &Node{data: data}
	node.next = l.head
	l.head = node

	if l.tail == nil {
		l.tail = l.head
	}

	l.size++
}

func (l *LinkedList) insertLast(data int) {
	if l.tail == nil {
		l.insertFirst(data)
		return
	}

	node := &Node{data: data}
	l.tail.next = node
	l.tail = node
	l.size++
}

func (l *LinkedList) insertAt(data int, index int) {
	if index < 0 || index > l.size {
		fmt.Println("Index out of bounds")
		return
	}

	if index == 0 {
		l.insertFirst(data)
		return
	}

	if index == l.size {
		l.insertLast(data)
		return
	}

	node := &Node{data: data}
	prev := l.head
	for i := 0; i < index-1; i++ {
		prev = prev.next
	}

	node.next = prev.next
	prev.next = node
	l.size++
}

func (l *LinkedList) deleteFirst() int {
	var value int = l.head.data
	if l.head == nil {
		return -1
	}

	l.head = l.head.next
	l.size--
	return value

}

func (l *LinkedList) deleteLast() int {
	var val int = l.tail.data
	if l.size <= 1 {
		l.deleteFirst()
	}
	var secondlast *Node = l.get(l.size - 2)
	l.tail = secondlast
	l.tail.next = nil

	l.size--

	return val

}

func (l *LinkedList) deleteAt(index int) int {
	if index < 0 || index >= l.size {
		fmt.Println("Index out of bounds")
		return -1
	}

	if index == 0 {
		return l.deleteFirst()
	}

	if index == l.size-1 {
		return l.deleteLast()
	}

	prev := l.get(index - 1)
	value := prev.next.data
	prev.next = prev.next.next
	l.size--
	return value
}

func (l *LinkedList) get(index int) *Node {
	temp := l.head
	for i := 0; i < index; i++ {
		temp = temp.next
	}

	return temp

}

func (l *LinkedList) printList() {

	node := l.head
	for node != nil {
		fmt.Print(node.data)
		fmt.Print("->")
		node = node.next
	}
	fmt.Println("nil")
}

func main() {
	ll := &LinkedList{}
	ll.insertFirst(1)
	ll.insertFirst(2)
	ll.insertFirst(3)
	ll.insertFirst(4)
	ll.insertFirst(5)
	ll.insertLast(6)
	ll.insertAt(7, 2)

	ll.deleteFirst()
	ll.deleteLast()
	ll.deleteAt(2)
	ll.printList()
}
